libraries/ESP8266WiFi/src/ESP8266WiFiMulti.cpp

```diff
@@ -1,267 +1,507 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ */
+
+#include "ESP8266WiFiMulti.h"
+#include <limits.h>
+#include <string.h>
+
+/**
+ * @brief Constructor
+ */
+ESP8266WiFiMulti::ESP8266WiFiMulti() : _firstRun(true)
+{
+}
+
+/**
+ * @brief Destructor
+ */
+ESP8266WiFiMulti::~ESP8266WiFiMulti()
+{
+    // Cleanup memory
+    APlistClean();
+}
+
+/**
+ * @brief Add Access Point
+ * @param ssid
+ *      WiFi SSID char array, max 32 characters + NULL character
+ * @param passphrase
+ *      WiFi password char array, max 63 characters + NULL character
+ * @retval true
+ *      Success
+ * @retval false
+ *      Failure
+ */
+bool ESP8266WiFiMulti::addAP(const char *ssid, const char *passphrase)
+{
+    return APlistAdd(ssid, passphrase);
+}
+
+/**
+ * @brief Remove all Access Points from list
+ */
+void ESP8266WiFiMulti::cleanAPlist(void)
+{
+    APlistClean();
+}
+
+/**
+ * @brief Check if Access Point exists in list
+ * @param ssid
+ *      WiFi SSID
+ * @param passphrase
+ *      WiFi Password
+ * @retval true
+ *      Success
+ * @retval false
+ *      Failure
+ */
+bool ESP8266WiFiMulti::existsAP(const char *ssid, const char *passphrase)
+{
+    return APlistExists(ssid, passphrase);
+}
+
+/**
+ * @brief Keep WiFi connected to Access Point with strongest WiFi signal (RSSI)
+ * @param connectTimeoutMs
+ *      Timeout in ms per WiFi connection (excluding fixed 5 seconds scan timeout)
+ * @return
+ *      WiFi status
+ */
+wl_status_t ESP8266WiFiMulti::run(uint32_t connectTimeoutMs)
+{
+    int8_t scanResult;
+    wl_status_t status;
+
+    // Fast connect to previous WiFi on startup
+    if (_firstRun) {
+        _firstRun = false;
+
+        // Check if previous WiFi connection saved
+        if (strlen(WiFi.SSID().c_str())) {
+            DEBUG_WIFI_MULTI("[WIFIM] Connecting saved WiFi\n");
+
+            // Connect to previous saved WiFi
+            WiFi.begin();
+
+            // Wait for status change
+            status = waitWiFiConnect(connectTimeoutMs);
+        }
+    }
+
+    // Check connection state
+    status = WiFi.status();
+    if (status == WL_CONNECTED) {
+        // Already connected
+        return status;
+    }
+
+    // Start WiFi scan
+    scanResult = startScan();
+    if (scanResult < 0) {
+        // No WiFi scan results
+        return WL_NO_SSID_AVAIL;
+    }
+
+    // Try to connect to multiple WiFi's with strongest signal (RSSI)
+    return connectWiFiMulti(connectTimeoutMs);
+}
+
+/**
+ * @brief Wait for WiFi connect status change, protected with timeout
+ * @param connectTimeoutMs
+ *      WiFi connection timeout in ms
+ * @return
+ *      WiFi connection status
+ */
+wl_status_t ESP8266WiFiMulti::waitWiFiConnect(uint32_t connectTimeoutMs)
```

Comment:
This method doesn't depend on anything from ESP8266WiFiMulti. Please remove it from the class and convert it to a standalone function that is private to the cpp (static or inside anonymous namespace)

Reply:
Accepted.